File: internal/mechanics/combat.go
package mechanics

import (
	"math"
	"math/rand"

	"github.com/slycrel/slycrel/internal/model"
)

// RollInitiative determines whether the player goes first.
// Original: RollInitiative() in Sly_TextCombatUnit.cp
// Returns true if the player has initiative.
func RollInitiative(playerMovement, monsterMovement int) bool {
	diff := playerMovement - monsterMovement
	return diff+rand.Intn(4) >= 0
}

// DoDamageUser calculates damage dealt by the player to a monster.
// Original: DoDamage(true) in Sly_TextCombatUnit.cp
func DoDamageUser(char *model.Character, monster *model.Monster) int {
	avgLevel := (char.ThiefLvl + char.MageLvl + char.FighterLvl) / 3

	var damage int
	if avgLevel <= 4 {
		// Low level formula
		low := int(math.Round(float64(char.Strength)*0.75+float64(char.Weapons[0].Strike)) * 1.25)
		high := int(math.Round((float64(char.Strength)/0.75 + float64(char.Weapons[0].Strike)) * 1.35))
		damage = RandBetween(low, high) - monster.Defense
		if damage < 2 {
			damage = rand.Intn(3) // 0-2
		}
	} else {
		// High level formula
		low := int(math.Round(float64(char.Strength)/2 + float64(char.Weapons[0].Strike)))
		high := int(math.Round((float64(char.Strength)/2 + float64(char.Weapons[0].Strike)) * 1.25))
		damage = RandBetween(low, high) - monster.Defense
	}

	if damage < 0 {
		damage = 0
	}
	return damage
}

// DoDamageMonster calculates damage dealt by a monster to the player.
// Original: DoDamage(false) in Sly_TextCombatUnit.cp
func DoDamageMonster(char *model.Character, monster *model.Monster) int {
	low := int(math.Trunc(float64(monster.Offense) * 0.35))
	high := int(math.Trunc(float64(monster.Offense) * 1.35))
	damage := RandBetween(low, high) - char.Armor[0].Defense

	if damage < 0 {
		damage = 0
	}
	return damage
}

// RandMonsterAction picks a random combat action for the monster.
// Original: rndMonsterAttack() in Sly_TextCombatUnit.cp
// Returns the action and whether the monster ran.
func RandMonsterAction(monsterHP, monsterMaxHP int) (model.CombatType, bool) {
	roll := RandBetween(1, 50)

	// Check if monster wants to run (low HP)
	wantsToRun := monsterMaxHP/20 > monsterHP

	switch {
	case roll <= 25:
		return model.CombatAttack, false
	case roll <= 30:
		return model.CombatParry, false
	case roll <= 37:
		return model.CombatBlock, false
	case roll <= 44:
		return model.CombatDodge, false
	default: // 45-50
		if wantsToRun {
			return model.CombatAttack, true // monster runs
		}
		return model.CombatAttack, false
	}
}

// CalcAdvanceMoves calculates how many extra attacks a combatant gets
// based on speed advantage in the combat loop.
// Original: advance moves calculation in ST_CalcAdvanceMoves
func CalcAdvanceMoves(attackerLoop, defenderLoop, combatLoop int) int {
	moves := 0
	loop := combatLoop
	for {
		loop--
		if loop <= 0 {
			break
		}
		if attackerLoop >= loop && defenderLoop < loop {
			moves++
		} else if (attackerLoop >= loop && defenderLoop >= loop) ||
			(attackerLoop < loop && defenderLoop >= loop) {
			break
		}
	}
	return moves
}

// ApplyUserAttackModifier adjusts damage based on the user's action vs monster's action.
// Original: modifier logic in CalcUserMove
func ApplyUserAttackModifier(baseDamage int, monMove model.CombatType) int {
	switch monMove {
	case model.CombatBlock:
		return baseDamage / 2
	case model.CombatParry:
		return int(math.Round(float64(baseDamage) * 0.6))
	default:
		return baseDamage
	}
}

// ApplyMonsterAttackModifier adjusts monster damage based on the player's action.
// Original: modifier logic in MonsterAttacks
func ApplyMonsterAttackModifier(baseDamage int, userMove model.CombatType, playerDex, monsterMove int) int {
	switch userMove {
	case model.CombatBlock:
		return baseDamage / 2
	case model.CombatParry:
		return int(math.Round(float64(baseDamage) * 0.6))
	case model.CombatSpecial:
		return int(math.Round(float64(baseDamage) * 1.75))
	case model.CombatDodge:
		if RandBetween(1, monsterMove) < playerDex {
			return 0
		}
		return baseDamage
	default:
		return baseDamage
	}
}

// Disposition represents the relative state of both combatants.
type Disposition int

const (
	DispMgoodUgood Disposition = iota + 1
	DispMgoodUbad
	DispMbadUgood
	DispMokUok
	DispMokUgood
	DispMokUbad
	DispMgoodUok
	DispMbadUok
	DispMbadUbad
)

// GetDisposition calculates the 9-state disposition based on HP percentages.
// Original: GetMainCombatText disposition calculation in Sly_TextCombatUnit.cp
func GetDisposition(userHP, userMax, monHP, monMax int) Disposition {
	var userPct, monPct int
	if userMax > 0 && userHP > 0 {
		userPct = int(math.Round(float64(userHP) / float64(userMax) * 100))
	}
	if monMax > 0 && monHP > 0 {
		monPct = int(math.Round(float64(monHP) / float64(monMax) * 100))
	}

	if userPct >= 70 {
		if monPct >= 70 {
			return DispMgoodUgood
		} else if monPct >= 35 {
			return DispMokUgood
		}
		return DispMbadUgood
	} else if userPct >= 35 {
		if monPct >= 70 {
			return DispMgoodUok
		} else if monPct >= 35 {
			return DispMokUok
		}
		return DispMbadUok
	} else {
		if monPct >= 70 {
			return DispMgoodUbad
		} else if monPct >= 35 {
			return DispMokUbad
		}
		return DispMbadUbad
	}
}

// RandBetween returns a random integer in [min, max] inclusive.
func RandBetween(min, max int) int {
	if min >= max {
		return min
	}
	return min + rand.Intn(max-min+1)
}

File: clients/web/mechanics/combat.js
import { CombatType } from '../model/enums.js';
import { randBetween } from './rand.js';

// Port of internal/mechanics/combat.go. Player vs monster damage rolls,
// initiative, monster AI, attack/defense modifiers, disposition flavor.

export function rollInitiative(playerMovement, monsterMovement) {
  const diff = playerMovement - monsterMovement;
  return diff + Math.floor(Math.random() * 4) >= 0;
}

export function doDamageUser(c, monster) {
  const avgLevel = Math.floor((c.thiefLvl + c.mageLvl + c.fighterLvl) / 3);
  const weapStrike = c.weapons[0].strike ?? 0;
  let damage;
  if (avgLevel <= 4) {
    const low = Math.round((c.strength * 0.75 + weapStrike) * 1.25);
    const high = Math.round((c.strength / 0.75 + weapStrike) * 1.35);
    damage = randBetween(low, high) - (monster.defense ?? 0);
    if (damage < 2) damage = Math.floor(Math.random() * 3);
  } else {
    const low = Math.round(c.strength / 2 + weapStrike);
    const high = Math.round((c.strength / 2 + weapStrike) * 1.25);
    damage = randBetween(low, high) - (monster.defense ?? 0);
  }
  return damage < 0 ? 0 : damage;
}

export function doDamageMonster(c, monster) {
  const off = monster.offense ?? 0;
  const low = Math.trunc(off * 0.35);
  const high = Math.trunc(off * 1.35);
  const armorDef = c.armor[0]?.defense ?? 0;
  const dmg = randBetween(low, high) - armorDef;
  return dmg < 0 ? 0 : dmg;
}

// Pick a random combat action for the monster. Returns { action, ran }.
export function randMonsterAction(monsterHP, monsterMaxHP) {
  const roll = randBetween(1, 50);
  const wantsToRun = Math.floor(monsterMaxHP / 20) > monsterHP;
  if (roll <= 25) return { action: CombatType.Attack, ran: false };
  if (roll <= 30) return { action: CombatType.Parry, ran: false };
  if (roll <= 37) return { action: CombatType.Block, ran: false };
  if (roll <= 44) return { action: CombatType.Dodge, ran: false };
  return { action: CombatType.Attack, ran: wantsToRun };
}

export function calcAdvanceMoves(attackerLoop, defenderLoop, combatLoop) {
  let moves = 0;
  let loop = combatLoop;
  while (true) {
    loop -= 1;
    if (loop <= 0) break;
    if (attackerLoop >= loop && defenderLoop < loop) {
      moves += 1;
    } else if ((attackerLoop >= loop && defenderLoop >= loop) ||
               (attackerLoop < loop && defenderLoop >= loop)) {
      break;
    }
  }
  return moves;
}

export function applyUserAttackModifier(baseDamage, monMove) {
  if (monMove === CombatType.Block) return Math.floor(baseDamage / 2);
  if (monMove === CombatType.Parry) return Math.round(baseDamage * 0.6);
  return baseDamage;
}

export function applyMonsterAttackModifier(baseDamage, userMove, playerDex, monsterMove) {
  if (userMove === CombatType.Block) return Math.floor(baseDamage / 2);
  if (userMove === CombatType.Parry) return Math.round(baseDamage * 0.6);
  if (userMove === CombatType.Special) return Math.round(baseDamage * 1.75);
  if (userMove === CombatType.Dodge) {
    return randBetween(1, monsterMove) < playerDex ? 0 : baseDamage;
  }
  return baseDamage;
}

// 9-state disposition based on HP%. 1=both good ... 9=both bad. Returns 0
// if either combatant has zero max HP (defensive).
export function getDisposition(userHP, userMax, monHP, monMax) {
  const userPct = userMax > 0 && userHP > 0 ? Math.round(userHP / userMax * 100) : 0;
  const monPct = monMax > 0 && monHP > 0 ? Math.round(monHP / monMax * 100) : 0;
  if (userPct >= 70) {
    if (monPct >= 70) return 1;       // MgoodUgood
    if (monPct >= 35) return 5;       // MokUgood
    return 3;                          // MbadUgood
  } else if (userPct >= 35) {
    if (monPct >= 70) return 7;       // MgoodUok
    if (monPct >= 35) return 4;       // MokUok
    return 8;                          // MbadUok
  } else {
    if (monPct >= 70) return 2;       // MgoodUbad
    if (monPct >= 35) return 6;       // MokUbad
    return 9;                          // MbadUbad
  }
}
